### PR TITLE
#19681 MalformedRequestContentRejection cause traceability

### PR DESCRIPTION
**Pull request by svezfaz**
- improved cause tracking for MalformedRequestContentRejection
- unwrapping RejectionErrors when unmarshalling
- amended old project name in CONTRIBUTING.md

### CODE REVIEW

**Comment by typesafehub-validator** [comment]
Hi @svezfaz,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Typesafe Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

File: akka-http/src/main/scala/akka/http/scaladsl/server/Rejection.scala [hunk 1]
```diff
@@ -99,7 +99,7 @@ case class TooManyRangesRejection(maxRanges: Int) extends Rejection
  * Note that semantic issues with the request content (e.g. because some parameter was out of range)
  * will usually trigger a `ValidationRejection` instead.
  */
-case class MalformedRequestContentRejection(message: String, cause: Option[Throwable] = None) extends Rejection
+case class MalformedRequestContentRejection(message: String, cause: Throwable) extends Rejection
```
Comment (patriknw) [inline akka-http/src/main/scala/akka/http/scaladsl/server/Rejection.scala]: this is not a binary compatible, I leave it to @ktoso to decide if we want to break it

**Comment by patriknw** [comment]
Refs #19681

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -59,7 +59,7 @@ The Akka build includes a special task called `validatePullRequest` which invest
 then running tests only on those projects.
 
 For example changing something in `akka-http-core` would cause tests to be run in all projects which depend on it
-(e.g. `akka-http-core-tests`, `akka-http-marshallers-*`, `akka-docs` etc.).
+(e.g. `akka-http-tests`, `akka-http-marshallers-*`, `akka-docs` etc.).
```
Comment (ktoso) [inline CONTRIBUTING.md]: thanks

File: akka-http/src/main/scala/akka/http/impl/server/RejectionHandlerWrapper.scala [hunk 3]
```diff
@@ -48,7 +48,7 @@ private[http] class RejectionHandlerWrapper(javaHandler: server.RejectionHandler
         case TooManyRangesRejection(maxRanges) ⇒
           handleTooManyRangesRejection(ctx, maxRanges)
         case MalformedRequestContentRejection(message, cause) ⇒
-          handleMalformedRequestContentRejection(ctx, message, cause.orNull)
+          handleMalformedRequestContentRejection(ctx, message, cause)
```
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/impl/server/RejectionHandlerWrapper.scala]: I see, ugly but we're removing this infra in any case now in new DSL

File: akka-http/src/main/scala/akka/http/scaladsl/server/Rejection.scala (hunk 1 above)
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/scaladsl/server/Rejection.scala]: I'd prefer to not change this one - don't see a strong reason to do so?

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/MarshallingDirectives.scala [hunk 4]
```diff
@@ -33,10 +33,11 @@ trait MarshallingDirectives {
       import ctx.materializer
       onComplete(um(ctx.request)) flatMap {
         case Success(value) ⇒ provide(value)
+        case Failure(RejectionError(r)) ⇒ reject(r)
         case Failure(Unmarshaller.NoContentException) ⇒ reject(RequestEntityExpectedRejection)
         case Failure(Unmarshaller.UnsupportedContentTypeException(x)) ⇒ reject(UnsupportedRequestContentTypeRejection(x))
         case Failure(x: IllegalArgumentException) ⇒ reject(ValidationRejection(x.getMessage.nullAsEmpty, Some(x)))
-        case Failure(x) ⇒ reject(MalformedRequestContentRejection(x.getMessage.nullAsEmpty, Option(x.getCause)))
+        case Failure(x) ⇒ reject(MalformedRequestContentRejection(x.getMessage.nullAsEmpty, x))
```
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/MarshallingDirectives.scala]: just `Option(x)` here, and keep `Option[Throwable]` in the rejection

**Comment by svezfaz** [comment]
Thanks @Konrad for reviewing. I assumed compatibility breakage was agreed in the issue discussion. I will amend this in the next days if that's OK.

**Comment by ktoso** [comment]
It seems you're right @svezfaz, I'm sorry for not having checked that discussion before reviewing, arguments there make sense! Seems then we can merge as is, thanks again :)

**Comment by ktoso** [comment]
OK TO TEST

**Comment by ktoso** [comment]
Merging and will add exception to rejection for better usability in case someone wants to inspect it

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/MarshallingDirectives.scala [hunk 5]
```diff
@@ -33,10 +33,11 @@ trait MarshallingDirectives {
       import ctx.materializer
       onComplete(um(ctx.request)) flatMap {
         case Success(value) ⇒ provide(value)
+        case Failure(RejectionError(r)) ⇒ reject(r)
```
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/MarshallingDirectives.scala]: 👍

**Comment by chadselph** [comment]
I can't seem to find the discussion about breaking source and binary compatibility, is it linked somewhere in here? I have mixed feelings, because I really needed this feature, but I got bit by the breaking change as well; so I'd like to understand when and why I can expect non-compatible changes in minor releases going forward.

**Comment by bfil** [comment]
Same here, the changes required were really minor but wouldn't have expected breaking changes in a patch release.

**Comment by ktoso** [comment]
Hey there,
Sorry this caught you off guard, however it is in-line with our binary compatibility rules as this is an _experimental_ module, as you'll notice by the word `-experimental` in the dependency (which we take very seriously and strictly enforce using tooling).

I noticed we don't have a single spot in Akka documentation where we explain it, the info is a bit spread around, so I'll fix that and add a documentation page.

The rules are described here:
- http://akka.io/news/2015/09/30/akka-2.4.0-released.html from 2.4 onwards our rules. Which is even more compatible than we used to be normally (2.4 is bin-compat with 2.3 which previously was not required – the next 2.5 will be binary compatible with 2.4 as well!)
- experimental rules http://doc.akka.io/docs/akka/snapshot/experimental/index.html (may break and disappear at will, that's how we're able to experiment with them)

Please notice that this is an experimental module (akka-http-core is stable, but akka-http is experimental still, exactly for that reason – that we anticipate breaking changes still).

**Comment by ktoso** [comment]
Will add docs, track here https://github.com/akka/akka/issues/20565

**Comment by ktoso** [comment]
wrote up the rules in this PR, hope I didn't mistake anything :) https://github.com/akka/akka/pull/20569 
Team will review in the morning; Sorry for not having this written up previously in one spot but scattered around the internets :)

**Pull request merged**